File: sdk/rust/examples/basic_usage.rs
//! Basic usage example for the Circuit Breaker Rust SDK
//!
//! This example demonstrates the core functionality of the SDK,
//! showing how to create workflows, resources, agents, and more.

use circuit_breaker_sdk::{
    create_agent, create_chat, create_resource, create_workflow,
    rules::{evaluate_rule, RuleBuilderStandalone},
    Client, PaginationOptions, Result, COMMON_MODELS,
};
use serde_json::json;

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    println!("🔧 Circuit Breaker Rust SDK Example");
    println!("===================================");

    // ============================================================================
    // 1. Create Client
    // ============================================================================

    println!("\n1. Creating client...");

    let mut client_builder = Client::builder()
        .base_url(
            &std::env::var("CIRCUIT_BREAKER_URL")
                .unwrap_or_else(|_| "http://localhost:4000".to_string()),
        )?
        .timeout(30000);

    if let Ok(api_key) = std::env::var("CIRCUIT_BREAKER_API_KEY") {
        client_builder = client_builder.api_key(api_key);
    }

    let client = client_builder.build()?;

    // Test connection
    match client.ping().await {
        Ok(response) => {
            println!("✅ Connected to Circuit Breaker v{}", response.version);

            match client.info().await {
                Ok(info) => {
                    println!("📊 Server: {}, Features: {:?}", info.name, info.features);
                }
                Err(_) => println!("📊 Server info not available"),
            }
        }
        Err(e) => {
            eprintln!("❌ Failed to connect: {}", e);
            return Err(e);
        }
    }

    // ============================================================================
    // 2. Create Workflow
    // ============================================================================

    println!("\n2. Creating workflow...");

    // Using the builder pattern
    let workflow_definition = create_workflow("Order Processing Example")
        .set_description("A simple order processing workflow")
        .add_state("pending", "normal")
        .add_state("validating", "normal")
        .add_state("processing", "normal")
        .add_state("completed", "final")
        .add_state("cancelled", "final")
        .add_transition("pending", "validating", "validate")
        .add_transition("validating", "processing", "approve")
        .add_transition("validating", "cancelled", "reject")
        .add_transition("processing", "completed", "complete")
        .add_transition("processing", "cancelled", "cancel")
        .set_initial_state("pending")
        .build();

    let workflow = client
        .workflows()
        .create()
        .name(workflow_definition.name)
        .description(workflow_definition.description.unwrap_or_default())
        .build()
        .await?;

    println!(
        "✅ Created workflow: {} ({})",
        workflow.name(),
        workflow.id()
    );

    // ============================================================================
    // 3. Create Resource
    // ============================================================================

    println!("\n3. Creating resource...");

    let resource_definition = create_resource(workflow.id())
        .add_data("orderId", json!("ORD-2024-001"))
        .add_data("customerId", json!("CUST-123"))
        .add_data("amount", json!(299.99))
        .add_data(
            "items",
            json!([
                {"id": "ITEM-1", "name": "Product A", "price": 199.99},
                {"id": "ITEM-2", "name": "Product B", "price": 100.0}
            ]),
        )
        .set_initial_state("pending")
        .build();

    let resource = client
        .resources()
        .create()
        .set_workflow_id(resource_definition.workflow_id)
        .add_data("orderId", json!("ORD-2024-001"))
        .add_data("customerId", json!("CUST-123"))
        .add_data("amount", json!(299.99))
        .set_initial_state("pending")
        .build()
        .await?;

    println!(
        "✅ Created resource: {} in state '{}'",
        resource.id(),
        resource.state().unwrap_or("unknown")
    );

    // ============================================================================
    // 4. Create Function (Skipped - handled via activities)
    // ============================================================================

    println!("\n4. Function creation (skipped - functions are handled via workflow activities)");

    // ============================================================================
    // 5. Server-Side Rule Storage & Evaluation
    // ============================================================================

    println!("\n5. Testing server-side rule storage...");

    // Try server-side rule creation and evaluation
    let rule_result = async {
        // Create a rule on the server (if available)
        let rule = client
            .rules()
            .create()
            .name("High Value Order Rule")
            .description("Detects orders over $1000 for approval workflow")
            .field_greater_than("amount", json!(1000))
            .add_tag("e-commerce")
            .add_tag("validation")
            .add_tag("high-value")
            .build()
            .await?;

        println!("✅ Created server rule: {} ({})", rule.name(), rule.id());

        // Test rule evaluation on the server
        let evaluation = client
            .rules()
            .evaluate(
                rule.id(),
                json!({
                    "orderId": "ORD-2024-001",
                    "customerId": "CUST-123",
                    "amount": 1500.0
                }),
            )
            .await?;

        let status = if evaluation.passed { "✅" } else { "❌" };
        println!("{} Server rule evaluation: {}", status, evaluation.reason);

        // List all rules
        let rules = client.rules().list().await?;
        println!("📋 Found {} rules on server:", rules.len());
        for rule in rules {
            println!(
                "  - {} ({}): {}",
                rule.name(),
                rule.id(),
                rule.description()
            );
        }

        Ok::<(), circuit_breaker_sdk::Error>(())
    }
    .await;

    if let Err(e) = rule_result {
        println!("⚠️ Server-side rules not available: {}", e);

        // Fallback to client-side rule evaluation
        println!("\n📋 Falling back to client-side rule evaluation...");

        let high_value_rule = RuleBuilderStandalone::field_greater_than(
            "high_value",
            "High value order",
            "amount",
            json!(1000),
        )
        .build();

        let resource_data = json!({
            "orderId": "ORD-2024-001",
            "customerId": "CUST-123",
            "amount": 299.99
        });

        let result = evaluate_rule(&high_value_rule, &resource_data);
        let status = if result.passed { "✅" } else { "❌" };
        println!("{} Client-side evaluation: {}", status, result.reason);
    }

    // ============================================================================
    // 6. Create Agent (if LLM is available)
    // ============================================================================

    println!("\n6. Creating agent...");

    let agent_result = async {
        let agent_definition = create_agent("Order Support Agent")
            .set_description("AI agent for customer order support")
            .set_type("conversational")
            .set_llm_provider("openai")
            .set_model(COMMON_MODELS::GPT_4O_MINI)
            .set_temperature(0.7)
            .set_max_tokens(500)
            .set_system_prompt(
                "You are a helpful customer service agent for an e-commerce platform. \
                Help customers with their orders, returns, and general inquiries. \
                Be polite, professional, and concise.",
            )
            .add_tool(
                "lookup_order",
                "Look up order details by order ID",
                json!({
                    "orderId": {"type": "string", "description": "The order ID to lookup"}
                }),
            )
            .set_memory("short_term", json!({"max_entries": 10}))
            .build();

        let agent = client
            .agents()
            .create()
            .name(agent_definition.name)
            .description(agent_definition.description.unwrap_or_default())
            .set_type("conversational")
            .set_llm_provider("openai")
            .set_model(COMMON_MODELS::GPT_4O_MINI)
            .set_temperature(0.7)
            .set_max_tokens(500)
            .set_system_prompt(
                "You are a helpful customer service agent for an e-commerce platform.",
            )
            .build()
            .await?;

        println!("✅ Created agent: {} ({})", agent.name(), agent.id());

        // Chat with the agent
        let chat_response = agent
            .send_message("Hi, I have a question about my order ORD-2024-001")
            .await?;
        println!("🤖 Agent response: {}", chat_response);

        Ok::<circuit_breaker_sdk::Agent, circuit_breaker_sdk::Error>(agent)
    }
    .await;

    match agent_result {
        Ok(_agent) => {
            println!("✅ Agent creation and testing successful");
        }
        Err(e) => {
            println!("⚠️ Agent creation skipped (LLM not available): {}", e);
        }
    }

    // ============================================================================
    // 7. Multi-Provider LLM Usage (if available)
    // ============================================================================

    println!("\n7. Multi-Provider LLM capabilities...");

    let llm_result = async {
        let llm_client = client.llm();

        // Test different providers
        println!("\n🧪 Testing multiple providers:");

        let test_models = vec![
            ("OpenAI GPT-4", COMMON_MODELS::GPT_4O_MINI),
            ("Claude Haiku", COMMON_MODELS::CLAUDE_3_HAIKU),
            ("Gemini Flash", COMMON_MODELS::GEMINI_FLASH),
        ];

        for (name, model) in test_models {
            let start_time = std::time::Instant::now();
            match llm_client
                .chat(
                    model,
                    "Explain workflow automation benefits in one sentence.",
                )
                .await
            {
                Ok(response) => {
                    let latency = start_time.elapsed().as_millis();
                    println!("✅ {} ({}ms): {}", name, latency, response);
                }
                Err(e) => {
                    println!("⚠️ {} unavailable: {}", name, e);
                }
            }
        }

        // Test virtual models for smart routing
        println!("\n🎯 Testing smart routing with virtual models:");

        let virtual_models = vec![
            ("Auto-Route", "auto"),
            ("Cost-Optimal", "cb:cost-optimal"),
            ("Fastest", "cb:fastest"),
        ];

        for (name, model) in virtual_models {
            match llm_client
                .chat(model, "What is the capital of France?")
                .await
            {
                Ok(response) => {
                    println!("🎯 {}: {}", name, response);
                }
                Err(e) => {
                    println!("⚠️ {} routing failed: {}", name, e);
                }
            }
        }

        // Cost comparison demonstration
        println!("\n💰 Provider cost comparison:");
        let cost_tests = vec![
            ("OpenAI", COMMON_MODELS::GPT_4O_MINI, 0.003),
            ("Anthropic", COMMON_MODELS::CLAUDE_3_HAIKU, 0.00025),
            ("Google", COMMON_MODELS::GEMINI_FLASH, 0.0000375),
        ];

        println!("Provider    | Model              | Est. Cost/1K tokens");
        println!("------------|--------------------|-----------------");

        for (provider, model, cost) in cost_tests {
            println!(
                "{:<11}| {:<18} | ${:.6}",
                provider,
                &model[..18.min(model.len())],
                cost
            );
        }

        // Using chat builder with different providers
        println!("\n🔧 Chat builder with provider selection:");
        let chat_builder = create_chat(COMMON_MODELS::CLAUDE_3_HAIKU)
            .set_system_prompt("You are a helpful assistant specializing in workflow automation.")
            .add_user_message("List 3 key benefits of workflow automation")
            .set_temperature(0.2)
            .set_max_tokens(150);

        let chat_result = chat_builder.execute(&llm_client).await?;
        if let Some(choice) = chat_result.choices.first() {
            println!("📝 Claude response: {}", choice.message.content);
        }

        Ok::<(), circuit_breaker_sdk::Error>(())
    }
    .await;

    if let Err(e) = llm_result {
        println!("⚠️ Multi-provider LLM testing skipped: {}", e);

        // Fallback to basic LLM usage
        let llm_client = client.llm();
        {
            match llm_client
                .chat(
                    COMMON_MODELS::GPT_4O_MINI,
                    "Explain the benefits of workflow automation in 2 sentences.",
                )
                .await
            {
                Ok(response) => {
                    println!("🤖 Fallback LLM response: {}", response);
                }
                Err(e) => {
                    println!("⚠️ All LLM usage skipped: {}", e);
                }
            }
        }
    }

    // ============================================================================
    // 8. Execute Workflow (Create Workflow Instance)
    // ============================================================================

    println!("\n8. Creating workflow instance...");

    let workflow_instance = workflow
        .execute_with_input(json!({
            "initialData": {
                "orderId": "ORD-2024-001",
                "customerId": "CUST-123",
                "amount": 299.99
            },
            "metadata": {
                "source": "api",
                "timestamp": chrono::Utc::now().to_rfc3339()
            }
        }))
        .await?;

    println!("✅ Created workflow instance: {}", workflow_instance.id());
    println!("📊 Current status: {:?}", workflow_instance.status());

    // Check instance status after a delay
    tokio::spawn(async move {
        tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
        // In a real implementation, you would refresh and check status
        println!("📊 Instance status check completed");
    });

    // ============================================================================
    // 9. Resource Operations
    // ============================================================================

    println!("\n9. Performing resource operations...");

    // Execute activity on resource using the workflow instance
    match client
        .resources()
        .execute_activity(
            resource.id().to_string(),
            "activity_0",
            Some(json!({
                "timestamp": chrono::Utc::now().to_rfc3339(),
                "performedBy": "sdk-user",
                "notes": "Executed validate activity"
            })),
        )
        .await
    {
        Ok(updated_resource) => {
            println!(
                "✅ Activity executed, resource state: {}",
                updated_resource.state().unwrap_or("unknown")
            );

            // Get resource history
            match client
                .resources()
                .get_history(
                    updated_resource.id().to_string(),
                    Some(PaginationOptions {
                        limit: Some(5),
                        offset: None,
                    }),
                )
                .await
            {
                Ok(history) => {
                    println!("📜 Resource history ({} events):", history.data.len());
                    for event in history.data {
                        println!(
                            "  - {}: {} → {} at {}",
                            event.activity,
                            event.from_state,
                            event.to_state,
                            event.timestamp.format("%H:%M:%S")
                        );
                    }
                }
                Err(e) => {
                    println!("⚠️ Could not fetch resource history: {}", e);
                }
            }

            // Update the workflow instance status
            println!(
                "📊 Instance update - State: {}",
                updated_resource.state().unwrap_or("unknown")
            );
        }
        Err(e) => {
            println!("⚠️ Activity execution failed: {}", e);
        }
    }

    println!("\n✨ Example completed successfully!");
    println!("\n🚀 For comprehensive multi-provider LLM testing, see:");
    println!("   cargo run --example multi_provider_demo");
    println!("\n📖 This demonstrates:");
    println!("   • Provider discovery and health monitoring");
    println!("   • Cost optimization across providers");
    println!("   • Real-time streaming from multiple providers");
    println!("   • Smart routing and virtual models");
    println!("   • Advanced features like function calling");

    Ok(())
}

File: sdk/typescript/examples/basic-usage.ts
/**
 * Basic usage example for Circuit Breaker TypeScript SDK
 *
 * This example demonstrates the core functionality of the simplified SDK,
 * showing how to create workflows, resources, functions, and agents.
 */

import {
  Client,
  createWorkflow,
  createResource,
  createAgent,
  createChat,
  COMMON_MODELS,
} from "../src/index.js";
import { RuleBuilder, evaluateRule } from "../src/rules.js";

async function main() {
  console.log("🔧 Circuit Breaker TypeScript SDK Example");
  console.log("=========================================");

  // ============================================================================
  // 1. Create Client
  // ============================================================================

  console.log("\n1. Creating client...");

  let clientBuilder = Client.builder()
    .baseUrl(process.env.CIRCUIT_BREAKER_URL || "http://localhost:4000")
    .timeout(30000);

  if (process.env.CIRCUIT_BREAKER_API_KEY) {
    clientBuilder = clientBuilder.apiKey(process.env.CIRCUIT_BREAKER_API_KEY);
  }

  const client = clientBuilder.build();

  // Alternative: Use the convenience SDK class
  // const sdk = createSDK({
  //   baseUrl: 'http://localhost:3000',
  //   apiKey: 'your-api-key'
  // });

  try {
    // Test connection
    const ping = await client.ping();
    console.log(`✅ Connected to Circuit Breaker v${ping.version}`);

    const info = await client.info();
    console.log(
      `📊 Server: ${info.name}, Features: ${info.features.join(", ")}`,
    );
  } catch (error) {
    console.error("❌ Failed to connect:", error);
    return;
  }

  // ============================================================================
  // 2. Create Workflow
  // ============================================================================

  console.log("\n2. Creating workflow...");

  try {
    // Using the builder pattern
    const workflowDefinition = createWorkflow("Order Processing Example")
      .setDescription("A simple order processing workflow")
      .addState("pending", "normal")
      .addState("validating", "normal")
      .addState("processing", "normal")
      .addState("completed", "final")
      .addState("cancelled", "final")
      .addTransition("pending", "validating", "validate")
      .addTransition("validating", "processing", "approve")
      .addTransition("validating", "cancelled", "reject")
      .addTransition("processing", "completed", "complete")
      .addTransition("processing", "cancelled", "cancel")
      .setInitialState("pending")
      .build();

    const workflow = await client.workflows().create(workflowDefinition);
    console.log(`✅ Created workflow: ${workflow.name} (${workflow.id})`);

    // ============================================================================
    // 3. Create Resource
    // ============================================================================

    console.log("\n3. Creating resource...");

    const resourceDefinition = createResource(workflow.id)
      .addData("orderId", "ORD-2024-001")
      .addData("customerId", "CUST-123")
      .addData("amount", 299.99)
      .addData("items", [
        { id: "ITEM-1", name: "Product A", price: 199.99 },
        { id: "ITEM-2", name: "Product B", price: 100.0 },
      ])
      .setInitialState("pending")
      .build();

    const resource = await client.resources().create(resourceDefinition);
    console.log(
      `✅ Created resource: ${resource.id} in state '${resource.state}'`,
    );

    // ============================================================================
    // 4. Create Function (Skipped - handled via activities)
    // ============================================================================

    console.log(
      "\n4. Function creation (skipped - functions are handled via workflow activities)",
    );

    // ============================================================================
    // 5. Server-Side Rule Storage & Evaluation
    // ============================================================================

    console.log("\n5. Testing server-side rule storage...");

    try {
      // Create a rule on the server
      const createRuleMutation = `
        mutation CreateRule($input: RuleInput!) {
          createRule(input: $input) {
            id
            name
            description
            version
            createdAt
            updatedAt
            tags
          }
        }
      `;

      const ruleInput = {
        name: "High Value Order Rule",
        description: "Detects orders over $1000 for approval workflow",
        condition: {
          conditionType: "FieldGreaterThan",
          field: "amount",
          value: 1000,
        },
        tags: ["e-commerce", "validation", "high-value"],
      };

      const createResult = (await client.mutation(createRuleMutation, {
        input: ruleInput,
      })) as any;

      if (createResult && createResult.createRule) {
        const createdRule = createResult.createRule;
        console.log(
          `✅ Created server rule: ${createdRule.name} (${createdRule.id})`,
        );

        // Test rule evaluation on the server
        const evaluateRuleMutation = `
          mutation EvaluateRule($input: RuleEvaluationInput!) {
            evaluateRule(input: $input) {
              ruleId
              passed
              reason
              details
            }
          }
        `;

        const evaluationInput = {
          ruleId: createdRule.id,
          data: {
            orderId: "ORD-2024-001",
            customerId: "CUST-123",
            amount: 1500.0,
          },
          metadata: {
            source: "api",
            timestamp: new Date().toISOString(),
          },
        };

        const evalResult = (await client.mutation(evaluateRuleMutation, {
          input: evaluationInput,
        })) as any;

        if (evalResult && evalResult.evaluateRule) {
          const evaluation = evalResult.evaluateRule;
          const status = evaluation.passed ? "✅" : "❌";
          console.log(`${status} Server rule evaluation: ${evaluation.reason}`);
        }

        // List all rules
        const listRulesQuery = `
          query ListRules {
            rules {
              id
              name
              description
              tags
              version
              createdAt
              updatedAt
            }
          }
        `;

        const listResult = (await client.query(listRulesQuery)) as any;
        if (listResult && listResult.rules) {
          console.log(`📋 Found ${listResult.rules.length} rules on server:`);
          listResult.rules.forEach((rule: any) => {
            console.log(
              `  - ${rule.name} (v${rule.version}): ${rule.description}`,
            );
            console.log(`    Tags: ${rule.tags.join(", ")}`);
            console.log(
              `    Created: ${new Date(rule.createdAt).toLocaleString()}`,
            );
          });
        }
      }
    } catch (error) {
      console.log(`⚠️ Server-side rules not available: ${error}`);

      // Fallback to client-side rule evaluation
      console.log("\n📋 Falling back to client-side rule evaluation...");

      const highValueRule = RuleBuilder.fieldGreaterThan(
        "high_value",
        "High value order",
        "amount",
        1000,
      );

      const resourceData = {
        orderId: "ORD-2024-001",
        customerId: "CUST-123",
        amount: 299.99,
      };

      const result = evaluateRule(highValueRule, resourceData);
      const status = result.passed ? "✅" : "❌";
      console.log(`${status} Client-side evaluation: ${result.reason}`);
    }

    // ============================================================================
    // 6. Create Agent (if LLM is available)
    // ============================================================================

    console.log("\n6. Creating agent...");

    try {
      const agentDefinition = createAgent("Order Support Agent")
        .setDescription("AI agent for customer order support")
        .setType("conversational")
        .setLLMProvider("openai")
        .setModel(COMMON_MODELS.GPT_O4_MINI)
        .setTemperature(0.7)
        .setMaxTokens(1000)
        .setSystemPrompt(
          `You are a helpful customer service agent for an e-commerce platform.
          Help customers with their orders, returns, and general inquiries.
          Be polite, professional, and concise.`,
        )
        .addTool("lookup_order", "Look up order details by order ID", {
          orderId: { type: "string", description: "The order ID to lookup" },
        })
        .setMemory("short_term", { max_entries: 10 })
        .build();

      const agent = await client.agents().create(agentDefinition);
      console.log(`✅ Created agent: ${agent.name} (${agent.id})`);

      // Chat with the agent
      const chatResponse = await client.agents().chat(agent.id, [
        {
          role: "user",
          content: "Hi, I have a question about my order ORD-2024-001",
        },
      ]);
      console.log(
        `🤖 Agent response:`,
        chatResponse.choices[0]?.message?.content,
      );
    } catch (error) {
      console.log(`⚠️  Agent creation skipped (LLM not available): ${error}`);
    }

    // ============================================================================
    // 7. Multi-Provider LLM Usage (if available)
    // ============================================================================

    console.log("\n7. Multi-Provider LLM capabilities...");

    try {
      const llmClient = client.llm();

      // Test different providers
      console.log("\n🧪 Testing multiple providers:");

      const testModels = [
        { name: "OpenAI GPT-4", model: COMMON_MODELS.GPT_O4_MINI },
        { name: "Claude Sonnet", model: COMMON_MODELS.CLAUDE_4_SONNET },
        { name: "Gemini Pro", model: COMMON_MODELS.GEMINI_PRO },
      ];

      for (const { name, model } of testModels) {
        try {
          const startTime = Date.now();
          const response = await llmClient.chat(
            model,
            "Explain workflow automation benefits in one sentence.",
            {
              temperature: 0.3,
              maxTokens: 1000,
            },
          );
          const latency = Date.now() - startTime;
          console.log(`✅ ${name} (${latency}ms): ${response}`);
        } catch (error) {
          console.log(`⚠️  ${name} unavailable: ${error}`);
        }
      }

      // Test virtual models for smart routing
      console.log("\n🎯 Testing smart routing with virtual models:");

      const virtualModels = [
        { name: "Auto-Route", model: "auto" },
        { name: "Cost-Optimal", model: "cb:cost-optimal" },
        { name: "Fastest", model: "cb:fastest" },
      ];

      for (const { name, model } of virtualModels) {
        try {
          const response = await llmClient.chat(
            model,
            "What is the capital of France?",
            { maxTokens: 1000 },
          );
          console.log(`🎯 ${name}: ${response}`);
        } catch (error) {
          console.log(`⚠️  ${name} routing failed: ${error}`);
        }
      }

      // Test streaming across providers
      console.log("\n🌊 Testing multi-provider streaming:");
      try {
        // Direct OpenAI API call for streaming demonstration
        const streamingResponse = await fetch(
          "http://localhost:3000/v1/chat/completions",
          {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
              model: COMMON_MODELS.GPT_O4_MINI,
              messages: [
                { role: "user", content: "Write a short haiku about AI" },
              ],
              stream: true,
              max_tokens: 1000,
            }),
          },
        );

        if (streamingResponse.ok && streamingResponse.body) {
          console.log("📝 Streaming response:");
          const reader = streamingResponse.body.getReader();
          const decoder = new TextDecoder();

          while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            const chunk = decoder.decode(value);
            const lines = chunk
              .split("\n")
              .filter((line) => line.trim() && line.includes("data:"));

            for (const line of lines) {
              if (line.includes("[DONE]")) break;
              try {
                const data = JSON.parse(line.replace("data: ", ""));
                const content = data.choices?.[0]?.delta?.content;
                if (content) process.stdout.write(content);
              } catch (e) {
                // Skip malformed JSON
              }
            }
          }
          console.log("\n✅ Streaming complete");
        } else {
          console.log("⚠️  Streaming not available");
        }
      } catch (error) {
        console.log(`⚠️  Streaming test failed: ${error}`);
      }

      // Cost comparison demonstration
      console.log("\n💰 Provider cost comparison:");
      const costTestPrompt = "Hello world";

      const costTests = [
        {
          provider: "OpenAI",
          model: COMMON_MODELS.GPT_O4_MINI,
          estimatedCost: 0.003,
        },
        {
          provider: "Anthropic",
          model: COMMON_MODELS.CLAUDE_4_SONNET,
          estimatedCost: 0.000025,
        },
        {
          provider: "Google",
          model: COMMON_MODELS.GEMINI_PRO,
          estimatedCost: 0.0000375,
        },
      ];

      console.log("Provider    | Model              | Est. Cost/1K tokens");
      console.log("------------|--------------------|-----------------");

      for (const test of costTests) {
        const providerPad = test.provider.padEnd(11);
        const modelPad = test.model.substring(0, 18).padEnd(18);
        const costStr = `$${test.estimatedCost.toFixed(6)}`;
        console.log(`${providerPad}| ${modelPad} | ${costStr}`);
      }

      // Using chat builder with different providers
      console.log("\n🔧 Chat builder with provider selection:");
      const chatBuilder = createChat(COMMON_MODELS.CLAUDE_4_SONNET)
        .setSystemPrompt(
          "You are a helpful assistant specializing in workflow automation.",
        )
        .addUserMessage("List 3 key benefits of workflow automation")
        .setTemperature(0.2)
        .setMaxTokens(150);

      const chatResult = await chatBuilder.execute(llmClient);
      console.log(
        `📝 Claude response:`,
        chatResult.choices[0]?.message?.content,
      );

      // Direct GraphQL llmChatCompletion example (matching Rust SDK approach)
      console.log("\n🔧 Direct GraphQL llmChatCompletion example:");
      const llmChatMutation = `
        mutation LlmChatCompletion($input: LlmchatCompletionInput!) {
          llmChatCompletion(input: $input) {
            id
            model
            choices {
              index
              message {
                role
                content
              }
              finishReason
            }
            usage {
              promptTokens
              completionTokens
              totalTokens
            }
          }
        }
      `;

      const llmInput = {
        model: COMMON_MODELS.GPT_O4_MINI,
        messages: [
          {
            role: "system",
            content:
              "You are a helpful assistant specializing in workflow automation.",
          },
          {
            role: "user",
            content:
              "What are the main advantages of using Circuit Breaker for LLM routing?",
          },
        ],
        temperature: 0.5,
        maxTokens: 1000,
      };

      try {
        const llmResult = (await client.mutation(llmChatMutation, {
          input: llmInput,
        })) as any;

        if (llmResult && llmResult.llmChatCompletion) {
          const response = llmResult.llmChatCompletion;
          console.log(
            `🚀 Direct GraphQL (${response.model}): ${response.choices[0]?.message?.content}`,
          );
          console.log(
            `   📊 Token usage: ${response.usage.totalTokens} total (${response.usage.promptTokens} + ${response.usage.completionTokens})`,
          );
        }
      } catch (graphqlError) {
        console.log(
          `⚠️  Direct GraphQL llmChatCompletion failed: ${graphqlError}`,
        );
      }
    } catch (error) {
      console.log(`⚠️  Multi-provider LLM testing skipped: ${error}`);

      // Fallback to basic LLM usage
      try {
        const llmClient = client.llm();
        const response = await llmClient.chat(
          COMMON_MODELS.GPT_O4_MINI,
          "Explain the benefits of workflow automation in 2 sentences.",
          {
            temperature: 0.3,
            maxTokens: 1000,
          },
        );
        console.log(`🤖 Fallback LLM response: ${response}`);
      } catch (fallbackError) {
        console.log(`⚠️  All LLM usage skipped: ${fallbackError}`);
      }
    }

    // ============================================================================
    // 8. Execute Workflow (Create Workflow Instance)
    // ============================================================================

    console.log("\n8. Creating workflow instance...");

    const workflowInstance = await client.workflows().execute(workflow.id, {
      initialData: {
        orderId: "ORD-2024-001",
        customerId: "CUST-123",
        amount: 299.99,
      },
      metadata: {
        source: "api",
        timestamp: new Date().toISOString(),
      },
    });
    console.log(`✅ Created workflow instance: ${workflowInstance.id}`);
    console.log(`📊 Current state: ${workflowInstance.state}`);

    // Check instance status
    setTimeout(async () => {
      try {
        const updatedInstance = await client
          .workflows()
          .getExecution(workflowInstance.id);
        console.log(`📊 Instance update - State: ${updatedInstance.state}`);
      } catch (error) {
        console.log(`⚠️  Could not fetch instance status: ${error}`);
      }
    }, 2000);

    // ============================================================================
    // 9. Resource Operations
    // ============================================================================

    console.log("\n9. Performing resource operations...");

    // Execute activity on resource using the workflow instance
    try {
      const activityResult = await client
        .resources()
        .executeActivity(workflowInstance.id, "activity_0", {
          data: {
            timestamp: new Date().toISOString(),
            performedBy: "sdk-user",
            notes: "Executed validate activity",
          },
        });
      console.log(
        `✅ Activity executed, resource state: ${activityResult.state}`,
      );

      // Get resource history
      const history = await client
        .resources()
        .getHistory(activityResult.id, { limit: 5 });
      console.log(`📜 Resource history (${history.data.length} events):`);
      history.data.forEach((event) => {
        console.log(
          `  - ${event.activity}: ${event.fromState} → ${event.toState} at ${new Date(event.timestamp).toLocaleTimeString()}`,
        );
      });
    } catch (error) {
      console.log(`⚠️  Activity execution failed: ${error}`);
    }

    console.log("\n✨ Example completed successfully!");
    console.log("\n🚀 For comprehensive multi-provider LLM testing, run:");
    console.log("   npx tsx examples/multi-provider-demo.ts");
    console.log("\n📖 This demonstrates:");
    console.log("   • Provider discovery and health monitoring");
    console.log("   • Cost optimization across providers");
    console.log("   • Real-time streaming from multiple providers");
    console.log("   • Smart routing and virtual models");
    console.log("   • Advanced features like function calling");
  } catch (error) {
    console.error("\n❌ Example failed:", error);

    if (error instanceof Error) {
      console.error("Error details:", {
        name: error.name,
        message: error.message,
        stack: error.stack?.split("\n").slice(0, 3).join("\n"),
      });
    }
  }
}

// Handle graceful shutdown
process.on("SIGINT", () => {
  console.log("\n👋 Shutting down gracefully...");
  process.exit(0);
});

process.on("unhandledRejection", (reason, promise) => {
  console.error("Unhandled Rejection at:", promise, "reason:", reason);
  process.exit(1);
});

// Run the example
if (require.main === module) {
  main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}

export { main };
